pypeit/io.py

```diff
@@ -286,3 +316,238 @@ def header_version_check(hdr, warning_only=True):
+
+    # If there aren't any arrays or tables, return an empty ImageHDU with just the header data.
+    if len(array_keys) < 1 and len(table_keys) < 1:
+        return fits.ImageHDU(header=hdr, name=name)
+
+    # If there's only a single array, return it in an ImageHDU
+    if len(array_keys) == 1:
+        return fits.ImageHDU(data=d[array_keys[0]], header=hdr,
+                             name=array_keys[0] if name is None else name)
+
+    # If there's only a single Table, return it in a BinTableHDU
+    if len(table_keys) == 1:
+        # TODO: If we pass hdr directly, does this call include any
+        # table meta?
+        return fits.BinTableHDU(data=d[table_keys[0]], header=hdr,
+                                name=table_keys[0] if name is None else name)
+
+    # Only remaining option is to build a BinTableHDU based on the
+    # dictionary contents.
+
+    # Do all arrays have the same number of rows?
+    single_row = len(numpy.unique([numpy.asarray(d[key]).shape[0] for key in array_keys])) > 1
+
+    # If the number of rows is inconsistent, save the data in a single
+    # row. Otherwise, save the data as a multi-row table.
+    cols = []
+    for key in array_keys:
+        cols += [fits.Column(name=key, format=rec_to_fits_type(d[key], single_row=single_row),
+                                 dim=rec_to_fits_col_dim(d[key], single_row=single_row),
+                                 array=numpy.expand_dims(d[key], 0) if single_row else d[key])]
+    return fits.BinTableHDU.from_columns(cols, header=hdr, name=name)
+
+
+def write_to_hdu(d, name=None):
+    """
+    Write the input to an astropy.io.fits HDU extension.
+
+    List and numpy.ndarray items are written as an ImageHDU,
+    `astropy.table.Table`_ items are written as a BinTableHDU, and
+    dictionaries are passed to :func:`dict_to_hdu`.
+    
+    .. warning::
+
+        If ``d`` is a list, the method assumes that the list is
+        essentially an array that can be sensibly converted to a
+        `numpy.ndarray`_.
+
+    Args:
+        d (:obj:`dict`, :obj:`list`, `numpy.ndarray`_, `astropy.table.Table`_):
+            Object to write to the HDU.
+        name (:obj:`str`, optional):
+            Name for the HDU extension.
+
+    Returns:
+        `astropy.io.fits.ImageHDU`_, `astropy.io.fits.BinTableHDU`_:
+        HDU with the data.
+
+    Raises:
+        TypeError:
+            Raised if the input object is not one of the allowed
+            types.
+
+    """
+    if isinstance(d, dict):
+        return dict_to_hdu(d, name=name)
+    if isinstance(d, Table):
+        return fits.BinTableHDU(data=d, name=name)
+    if isinstance(d, (numpy.ndarray, list)):
+        return fits.ImageHDU(data=numpy.asarray(d), name=name)
+    raise TypeError('Input must be a dictionary, astropy.table.Table, list, or numpy.ndarray.')
+
+
+def write_to_fits(d, ofile, name=None, overwrite=False, checksum=True):
```

Comment:
Should this replace `pypeit.core.save.write_fits()`?

Also, I think `hdr` for the Primary HDU should be an optional argument and should default to `pypeit.io.initialize_header()` if not provided.

Reply:
For the first question, yes, I think so. This function is more general, but with a little more overhead, as well...

Good idea.  I've added `hdr` as an input option.  I've gone ahead and used `initialize_header` to set the default, but the function is more general than that.